### ISSUE TITLE
"Ignore_row_on_dupkey_index" hint on INSERT triggers NJS-027

### ISSUE THREAD

**Issue by sanfords** [oracle/node-oracledb#343]
Hello,

  We have a table.  This table has a few columns.  One of them is `string_1`.  `string_1` has a unique constraint.  It was added to the table with:

```
ALTER TABLE unittest ADD (blob_1 blob,uint32_1 number,date_1 timestamp(5),string_1 varchar2(250),CONSTRAINT string_1_uk UNIQUE (string_1))
```

  When we do our first insert into this table it works fine, as in:

```
INSERT /*+ ignore_row_on_dupkey_index(unittest, string_1_uk) */ INTO UNITTEST (STRING_1,SPOC_CM_ID) VALUES (:string_1,:spoc_cm_id) RETURNING ID INTO :id

values: { id: { type: 2002, dir: 3003 }, string_1: '0', spoc_cm_id: -1 }
```

  However when we do our _next_ insert with the exact same SQL and value object, we get the following error:

```
[Error: NJS-027: unexpected SQL parsing error] 
```

  We're in a bind at the moment.  Any ideas why this is happening?

Thank you!

PS - `id` is an auto incrementing id set with a trigger _before_ insert with:

```
create or replace trigger unittestrbi before insert on unittest referencing old as old new as new   for each row     begin       :new.id := unittest_seq.nextval;     end; 
```

**Comment by sanfords** [comment]
I wanted to followup with another test.  We've tried an alternate method for detecting uniqueness:

```
sql DECLARE ROW_COUNT NUMBER; BEGIN SELECT COUNT(*) INTO ROW_COUNT FROM UNITTEST  where (string_1 like :string_1); IF (ROW_COUNT = 0) THEN INSERT INTO UNITTEST (BLOB_1,STRING_1,SPOC_CM_ID) VALUES (EMPTY_BLOB(),:string_1,:spoc_cm_id) RETURNING ID,BLOB_1 INTO :id,:blob_1; END IF; END;
 ```

This one returns:

```
[Error: DPI-007: invalid OCI handle or descriptor]
```

We're kind of stuck.

**Comment by sanfords** [comment]
One more addition (last one!)... if we don't do any of the above, and we rely solely on the return unique constraint error (-00001), our inserts start stalling and slowing our ingest process dramatically.

**Comment by cjbj** [comment]
@sanfords the NJS-027 was added to work around a case where the DB didn't correctly identify a syntax error.  I forget whether it was the issue you reported https://github.com/oracle/node-oracledb/issues/203.

Better post a full test case including DDL to create the schema.  And give us the platform, version and Oracle client & server version.  Thanks.

**Comment by sanfords** [comment]
Wow, it's official, I'm losing my mind!  :-)

I will roll an example for you.  Thanks for the quick response.

**Comment by sanfords** [comment]
Sorry for the delay...

-- Verions:

v1.5 node-oracledb on Mac OS X 10.11 and Linux.  (Reproduces the same on both.)
instantclient-basic-macos.x64-11.2.0.4.0.zip
instantclient-sdk-macos.x64-11.2.0.4.0.zip
Oracle DB 12.1.0.2.0 on x86_64 Linux
Oracle Linux Server release 5.10

```
-- using 'hint' to ignore duplicate and return.

create sequence unittest_seq increment by 1 start with 1 nomaxvalue cache 50 order;
create table unittest ( id number not null primary key, spoc_cm_id number, created_timestamp timestamp(5) default systimestamp, modified_timestamp timestamp(5) default systimestamp );
create or replace trigger unittestrbi before insert on unittest referencing old as old new as new   for each row     begin       :new.id := unittest_seq.nextval;     end;
create or replace trigger unittestrbu before update on unittest referencing old as old new as new   for each row     begin       :new.modified_timestamp := systimestamp;     end;  
ALTER TABLE unittest ADD (blob_1 blob,uint32_1 number,date_1 timestamp(5),string_1 varchar2(250),CONSTRAINT string_1_uk UNIQUE (string_1))

-- ...and then...
INSERT /*+ ignore_row_on_dupkey_index(unittest, string_1_uk) */ INTO UNITTEST (STRING_1,SPOC_CM_ID) VALUES (:string_1,:spoc_cm_id) RETURNING ID INTO :id
  values: { id: { type: 2002, dir: 3003 }, string_1: '0', spoc_cm_id: -1 }

-> [Error: NJS-027: unexpected SQL parsing error]
```

Thank you for the help!

**Comment by sanfords** [comment]
(We're experimenting with an IF/ELSE to check for count equal to 0 before insert, but it's being temperamental.  I think that one was our fault.)

**Comment by cjbj** [comment]
I can reproduce it.  We'll have to debug it.

**Comment by cjbj** [comment]
@sanfords can you see if this patch to `src/njs/src/njsConnection.cpp` resolves this:
```
@@ -1498,7 +1498,8 @@ void Connection::Async_Execute (uv_work_t *req)
       executeBaton->rowsAffected = executeBaton->dpistmt->rowsAffected();
 
       // Check whether indicators were allocated as part of callback
-      if ( executeBaton->stmtIsReturning )
+      // Address GitHub issue #343
+      if ( executeBaton->stmtIsReturning && executeBaton->rowsAffected )
       {
         for ( unsigned int b = 0; b < executeBaton->binds.size (); b++ )
         {
```

I'll update your other issue #344 with a second patch.

**Comment by sanfords** [comment]
This one appears to be fixed!

However we would prefer the speed and fidelity of this one: https://github.com/oracle/node-oracledb/issues/344#issuecomment-182149037

**Comment by cjbj** [comment]
This is fixed in 1.7
